x-pack/metricbeat/module/aws/awshealth/awshealth.go

```diff
@@ -0,0 +1,436 @@
+				if !ok {
+					return nil
+				}
+				m.Logger().Debugf("[AWS Health] [DescribeEventDetails] Event ARN : %s, Affected Entities (Pending) : %d, Affected Entities (Resolved): %d, Affected Entities (Others) : %d", *healthDetails.event.Arn, healthDetails.affectedEntityPending, healthDetails.affectedEntityResolved, healthDetails.affectedEntityOthers)
+				events = append(events, createEvents(healthDetails))
+			}
+		}
+		wg.Wait()
+		if eventOutput.NextToken == nil {
+			break
+		} else {
+			nextTokenString = *eventOutput.NextToken
+		}
+	}
+	close(c)
+	close(errCh)
+	return events
+}
+
+func createEvents(hd HealthDetails) mb.Event {
+	currentDate := getCurrentDateTime()
+	eventID := currentDate + getStringValueOrDefault(hd.event.Arn) + getStringValueOrDefault((*string)(&hd.event.StatusCode))
+	event := mb.Event{
+		MetricSetFields: mapstr.M{
+			"event_arn":                  getStringValueOrDefault(hd.event.Arn),
+			"end_time":                   getTimeValueOrDefault(hd.event.EndTime),
+			"event_scope_code":           getStringValueOrDefault((*string)(&hd.event.EventScopeCode)),
+			"event_type_category":        getStringValueOrDefault((*string)(&hd.event.EventTypeCategory)),
+			"event_type_code":            getStringValueOrDefault(hd.event.EventTypeCode),
+			"last_updated_time":          getTimeValueOrDefault(hd.event.LastUpdatedTime),
+			"region":                     getStringValueOrDefault(hd.event.Region),
+			"service":                    getStringValueOrDefault(hd.event.Service),
+			"start_time":                 getTimeValueOrDefault(hd.event.StartTime),
+			"status_code":                getStringValueOrDefault((*string)(&hd.event.StatusCode)),
+			"affected_entities_pending":  hd.affectedEntityPending,
+			"affected_entities_resolved": hd.affectedEntityResolved,
+			"affected_entities_others":   hd.affectedEntityOthers,
+			"affected_entities":          createAffectedEntityDetails(hd.affectedEntities),
+		},
+		RootFields: mapstr.M{
+			"cloud.provider": "aws",
+		},
+		ID: generateEventID(eventID),
+	}
+	return event
+}
+
+type HealthDetails struct {
+	event                  types.Event
+	eventDescription       string
+	affectedEntities       []types.AffectedEntity
+	affectedEntityPending  int32
+	affectedEntityResolved int32
+	affectedEntityOthers   int32
+}
+
+type AffectedEntityDetails struct {
+	AwsAccountId    string    `json:"aws_account_id"`
+	EntityUrl       string    `json:"entity_url"`
+	EntityValue     string    `json:"entity_value"`
+	LastUpdatedTime time.Time `json:"last_updated_time"`
+	StatusCode      string    `json:"status_code"`
+	EntityArn       string    `json:"entity_arn"`
+}
+
+// getStringValueOrDefault returns the string value or an empty string if the pointer is nil.
+func getStringValueOrDefault(s *string) string {
+	if s != nil {
+		return *s
+	}
+	return ""
+}
+
+func getTimeValueOrDefault(t *time.Time) time.Time {
+	if t != nil {
+		return *t
+	}
+	return time.Time{}
+}
```

Comment:
I suggested a single generic function for this. Did you have a look? It is in the diff patch.

Reply:
Addressed